### ISSUE TITLE
Inaccurate tone mapping

### ISSUE THREAD

**Issue by movieperson** [haasn/libplacebo#175]
I'm having issues with color and saturation inaccuracies mapping HDR to both BT.709 and BT.2020.  I am attaching some DaVinci analyses of these captures for you to have a look at.  Two basic issues I see are premature clipping of the red channel and inaccuracies in the color target points.   have a FP with a 110 nit white level.  I have tried mapping to both BT-709 and BT-2020 and both environments show the same red clipping and primary color target shifts.

target-peak=100
![MPV 100](https://github.com/haasn/libplacebo/assets/72896057/d0c47767-fac4-449f-84b0-1ca0465dfa0c)

target-peak=300
![MPV 300](https://github.com/haasn/libplacebo/assets/72896057/84561e9f-ab44-478a-a83c-2e3daa5b4a0c)

target-peak=500
![MPV 500](https://github.com/haasn/libplacebo/assets/72896057/cd1dd4cf-79ea-49d2-b182-c5161314227b)

target-peak=700
![MPV 700](https://github.com/haasn/libplacebo/assets/72896057/6cff7422-b40e-4ffd-9a81-c3ceb7ad377e)

target-peak=900
![MPV 900](https://github.com/haasn/libplacebo/assets/72896057/62570a64-808f-4632-930e-92f326a7ddf2)

target-peak=1000
![MPV 1000](https://github.com/haasn/libplacebo/assets/72896057/30f74547-5e31-469b-a19e-68af0d8eaf52)

The issue changes related to the target-peak settings and is shown very clearly in these two captures.  Notice that the floor and the chair cushion are blown out in the reds at lower target peak values.  

target-peak=500
![Screenshot 500](https://github.com/haasn/libplacebo/assets/72896057/b75084bb-e010-48d3-81f7-2e1db01a5aac)

target-peak=100
![Screenshot 100 ](https://github.com/haasn/libplacebo/assets/72896057/9af059ec-c118-4e0a-b47a-c7663332c9f6)

I have tried multiple renderers, and settings, but the issue seems related to the tone mapping.  target-peak=disabled does nothing to help this.  You can check out a thread here to look at the entire discussion: [https://www.avsforum.com/threads/unofficial-mpv-player-support-thread.3274072/page-11#post-62631336](url)

I appreciate you having a look at this.  Thanks, Vern Dias

**Comment by pbiancardi** [comment]
Additionally no combination of settings fixes the red issues shown in the pictures above if target-peak is set low (say under 100).

Current questions are -

1. Are people with 50 nit projectors supposed to use target-peak=50 (which results in that horrible red mess in the second picture) or leave target-peak=auto which seems wrong since actual target-peak is much less than 203 (but this would look more correct on screen).

2. Is this something that the new gamut-mapping options will be addressing?

**Comment by nicko88** [comment]
> 1. Are people with 50 nit projectors supposed to use target-peak=50 (which results in that horrible red mess in the second picture) or leave target-peak=auto which seems wrong since actual target-peak is much less than 203 (but this would look more correct on screen).

This is something I am looking to find an answer for as well.

What are people with ~50 nit display's supposed to use for settings to get a good quality output?

**Comment by haasn** [comment]
I think that you're probably better off leaving target-peak at 203 for projectors, since they were intended to map SDR material. It will obviously be darker than an equivalent HDR projector, but then again, an equivalent HDR projector wouldn't be limited to 50 nits.

**Comment by nicko88** [comment]
I feel like this is not the right decision personally.

madVR looks excellent at 50 nits, so I don't see why libplacebo couldn't as well.  I would like to be able to some day replace madVR with libplacebo, but as it stands right now, at 50 nits or even 75 nits it doesn't come close to the look of madVR at low nits.

SDR looks fine at 50 nits, and most projector calibrations for SDR are done at 50 nits and have been for a very long time.

There is no such thing really as an "HDR projector".  All projectors have a single backlight and no local dimming ability and I doubt this will change any time soon.  The peak brightness you get of course depends on your screen size and gain, and so there are lots of people who have modern projectors that support the HDR format, but they get well under 100 nits due to screen size and gain setup.

I don't think we should necessarily have to suffer a dim and dull image when playing HDR sourced content to get the 4K resolution, wide color, and most importantly the modern remastered video which holds a lot more quality than the old 1080p BluRay release does.  Another factor is for streaming sourced video, the bitrate of the 4K HDR is much higher and thus is clearer and sharper than the SDR stream.

I took some shots here of how my madVR setup outputs, vs how the latest libplacebo outputs at the default 203 nits.

https://nicko88.com/misc/compare/madVR-mpv/greatest_showman/

When viewed on my modern 4K JVC projector at ~60 nits output, the madVR looks significantly better and enjoyable, and the mpv as expected looks quite dim, dull, and flat.  Hard to see any shadow details, etc.

I personally think that when using a display like a projector with such a low contrast and no local dimming, that the best image is achieved when making the image utilize the entire dynamic range of the image itself, rather than reserving so much of it for the occasional HDR highlight.  You are giving up a lot of the display's native contrast most of the time for a few highlights.

This is not just my own opinion, but when I host and go to get togethers with groups of HT enthusiasts, the overwhelming response from people viewing both ways is they much prefer how the projectors look when it's tone-mapped brighter.


I hope this is not coming across wrong.  I just hope and wish that some other tone-mapping solution besides madVR and Lumagen would care about improving and maximizing the performance on very commonly less than 100 nit projector setups.

**Comment by haasn** [comment]
Maybe we should just reduce the saturation considerably when dramatically lowering brightness.

Can you throw me a hevc still frame to reproduce your tests? There are some obvious things we could try doing (e.g. just literally scaling down chromaticity when reducing brightness)

**Comment by nicko88** [comment]
Yes, I will look at getting you HEVC still frames.

Do you know an appropriate ffmpeg command to get such an image?  I am not 100% sure of the correct command to get that.

Otherwise I can also just lossless clip tiny video chunks.

**Comment by haasn** [comment]
For example, here's a simple diff:

```diff
diff --git a/src/shaders/colorspace.c b/src/shaders/colorspace.c
index ddbd663e..213c96cc 100644
--- a/src/shaders/colorspace.c
+++ b/src/shaders/colorspace.c
@@ -1866,8 +1866,8 @@ void pl_shader_color_map_ex(pl_shader sh, const struct pl_color_map_params *para
             GLSL("ipt.x = tone_map(ipt.x); \n");
         }
 
-        // Avoid raising saturation excessively when raising brightness
-        GLSL("ipt.yz *= min(i_orig / ipt.x, 1.0); \n");
+        // Avoid raising saturation excessively when changing brightness
+        GLSL("ipt.yz *= min(i_orig / ipt.x, ipt.x / i_orig); \n");
 
         if (hybrid_mix > 0) {
             GLSL("vec3 lmsclip = lmspq;                     \n"
```

**Comment by haasn** [comment]
> Do you know an appropriate ffmpeg command to get such an image? I am not 100% sure of the correct command to get that.

```
ffmpeg -ss HH:MM:SS -i INPUT -vframes 1 -vcodec copy -an out.hevc
```

**Comment by nicko88** [comment]
> Can you throw me a hevc still frame to reproduce your tests?

Here are all the frames from my previous comparison images.

[frames.zip](https://github.com/haasn/libplacebo/files/11877287/frames.zip)

**Comment by damian101** [comment]
Whenever I messed around with libplacebo tonemapping in the past, I felt like something was fundamentally wrong, especially since messing with things like target peak really did not produce the results I expected. But now, after extensive testing across different libplacebo implementations, I am certain that libplacebo tonemapping is just plain incorrect, probably across all tonemapping functions. It can be easily observed by using the simple tonemapping function clip, no matter how low I set the target peak, I could not observe additional clipping, pixels way brighter than the specified target peak did not clip at all, and source peak also had an influence on clip tonemapping, which it should not. It can also be easily observed that mobius tonemapping is incorrect, dark areas in the low linear, absolute part of the funciton, should be completely independent of the specififed source peak, which is not the case. Generally libplacebo tonemapping seems to be always done somehow relative to the specified source peak, which should not be the case. Please fix.

**Comment by haasn** [comment]
> It can be easily observed by using the simple tonemapping function clip, no matter how low I set the target peak, I could not observe additional clipping, pixels way brighter than the specified target peak did not clip at all, and source peak also had an influence on clip tonemapping, which it should not.

It seems to work fine for me from a quick test. Without a better description of your methodology I cannot help you.

**Comment by pbiancardi** [comment]
> @nicko88 seems the patch I posted earlier helps a lot, I will merge it
> 
> Apart from that, I found that increasing both 'Contrast recovery' (to about 0.7) and 'Contrast' (for spline, to about 0.9) gave me a look&feel similar to madVR, just testing on the first few screenshots:
> 
> ![libplacebo-50-mod-custom](https://user-images.githubusercontent.com/1149047/249101334-2106f573-1643-4fad-b021-8ed23334b5e6.png) ![libplacebo3](https://user-images.githubusercontent.com/1149047/249102162-ca5a0a9a-2864-4567-b5d7-6595f5fa28e7.png)

Are contrast and contrast recovery set in mpv.conf for MPV or how do we access these settings?  Appreciate the work on this.

**Comment by haasn** [comment]
Spline contrast is exposed as `--tone-mapping-param`, contrast recovery is not exposed as an option yet; will add it soon.

As a side note, the above change and a whole pile of other related tone mapping improvements are sitting on a branch at https://code.videolan.org/videolan/libplacebo/-/merge_requests/481

I'll merge that and make the libplacebo release before updating mpv

**Comment by movieperson** [comment]
Here are some of my problem frames in HEVC format.  Sorry, if they are a little late.

[HEVC.zip](https://github.com/haasn/libplacebo/files/11882732/HEVC.zip)

They all have red push / clipping issues with MPV and Jriver, while looking good with MadVR.

Thanks again for looking at this.

**Comment by pbiancardi** [comment]
Here is one I watched last night, curious how this looks after the new updates.

Source - https://www.dropbox.com/s/ziijjf5qxiy8bi7/In%20the%20line%20of%20fire.mkv?dl=0

<img width="960" alt="Side By Side" src="https://github.com/haasn/libplacebo/assets/133981792/0b765770-ed99-40d8-a097-d961f047c9a4">

**Comment by damian101** [comment]
> > It can be easily observed by using the simple tonemapping function clip, no matter how low I set the target peak, I could not observe additional clipping, pixels way brighter than the specified target peak did not clip at all, and source peak also had an influence on clip tonemapping, which it should not.
> 
> It seems to work fine for me from a quick test. Without a better description of your methodology I cannot help you.

Just compare vo=gpu tonemapping (which is correct as far as I can see, and also does the same as the vast majority of other simple tonemapping implementations, different desaturion methods for out-of-range highlights aside) with vo=gpu-next tonemapping.

First use a simple mpv config like this on any HDR source:
```
vo=gpu
target-prim=bt.709
target-trc=bt.1886
gamut-mapping-mode=clip
tone-mapping=clip
tone-mapping-mode=rgb
dither-depth=auto
target-peak=10
```
and mess around with target-peak. You'll see that this "clip" tonemapping function will just set the SDR output peak to the specified target-peak and linearly map all values from 0 to it to that range, while simply clipping all values above it (implementations are different due to the nature of pixels being made of three sub-pixels of often very different intensity, so desaturation is usually applied to out-of-range highlights, to help them clip to white quicker, improving relative brightness accuracy at the expense of color accuracy). If you set the target peak low enough (10 seems to be the lower limit for mpv for some reason, you will get severe clipping on most content. Now switch to vo=gpu-next without changing any of the other settings. The output will be completely different, zero (additional) clipping. Target-peak won't even have an affect until you set it higher than the source peak. This is just wildly inaccurate behavior for tonemapping function "clip", more something I would expect from "linear".
Here an example of the drastic difference between correct clip tonemapping and libplacebo (target peak 10 nits):
![mpv-shot0006](https://github.com/haasn/libplacebo/assets/37085490/66dd8846-0a89-435c-8b11-4f031f468fbb)
![mpv-shot0007](https://github.com/haasn/libplacebo/assets/37085490/dc4c7b81-6bed-44f1-b10f-67cff1e99c00)

And it's not only clip that is broken, the others are too. Mobius for example is supposed to have an absolute linear part at the bottom, and have the rest smoothly mapped into the target range. This lower linear part of the function is simple absolute 1:1 tonemapping, identical to simple clip. But while for implementations like mpv vo=gpu and and vapoursynth tonemap plugin this is true, it is not for libplacebo, very dark regions there are completely different from the expected 1:1 tonemapping, and generally doesn't behave at all like I'm used to from other implementations.

**Comment by movieperson** [comment]
> I think that you're probably better off leaving target-peak at 203 for projectors, since they were intended to map SDR material. It will obviously be darker than an equivalent HDR projector, but then again, an equivalent HDR projector wouldn't be limited to 50 nits.

I  believe you may be making an incorrect assumption here.  The JVC RS2100, 3100, and 4100 are considered to be HDR capable projectors, supporting the DCI P3 and BT.2020 color spaces (more or less, not having full gamut coverage).  Unless you move up to something like a Christie, which requires a 220v power source, you are not going to get high nit levels in an HT  

Even a commercial theatre is unlikely to get 200 nits on the screen, though there are exceptions, usually running a dual projector installation, which most theatres can't even afford to purchase.

Yet, on the JVC, no matter what color space you chose, the maximum nits on the screen are only dependent on the the laser level you choose to run at.  

For example I get 64 nits (19fl) on low, 103 nits (30fl)on medium, and 142 nits (42fl) on high on a 52"x144" screen.  Granted my seating distance is extremely close (~8') and my room is a bat cave with no external light sources, but with high quality tone mapping an HDR image can be almost uncomfortably bright at times.

Also, the absolute white level is only a part of the viewing experience.  It's really the overall contrast of the image that counts, and in that respect the JVC excels with a roughly 50,000:1 contrast.

In conclusion, since the market seems to be projection with relatively low light output in a light controlled room or emissive panels with a higher light output, but likely used in a non light controlled room, it seems to be a prudent choice to provide optimum mapping solutions for both of these environments.

**Comment by kasper93** [comment]
@damian101 minimum target-peak value for libplacebo is 50.

**Comment by damian101** [comment]
> @damian101 minimum target-peak value for libplacebo is 50.

That might be.
Comparison at 50 nits shows the same issue though, proper clipping at 50 nits for mpv vo=gpu, but seemingly the full source dynamic range (not very large, global peak is 457 nits for this source, and source peak used here is even lower thanks to per-scene Dolby Vision metadata, that's why the libplacebo result looks quite decent) is preserved for the libplacebo equivalent:
![mpv-shot0010](https://github.com/haasn/libplacebo/assets/37085490/be8b1ebc-eb7e-4f92-84ab-820a269f27ce)
![mpv-shot0011](https://github.com/haasn/libplacebo/assets/37085490/870dd234-0863-4af2-9ead-8d00b9722f46)

**Comment by damian101** [comment]
> @damian101 minimum target-peak value for libplacebo is 50.

Oh... Then it does actually clip. But I think it's still incorrect, the tonemapping doesn't appear to be neutral, dark areas are too dark.
Here both at 50 nits:
![mpv-shot0014](https://github.com/haasn/libplacebo/assets/37085490/0f58bf2b-a34d-4676-9b25-f43e462c1f71)
![mpv-shot0015](https://github.com/haasn/libplacebo/assets/37085490/1ca62b3a-d5ba-405b-aa95-465727475e7b)

This is not a matter of aesthetics, the target transfer curve together with the target peak allow every pixel with in target range brightness to be mapped neutrally, there shouldn't be differences between implementations of the "clip" tonemapping function, problems with colors, subpixels aside.

**Comment by kasper93** [comment]
> Oh... Then it does actually clip. But I think it's still incorrect, the tonemapping doesn't appear to be neutral, dark areas are too dark.

And why you think it is incorrect? You expect to clip and you are clipping. In both peak brightness and black point. Without ICC profile, 1000 contrast is assumed so black point would be equal to target_peak / 1000 nits. vo_gpu limitation is that it doesn't have BPC, it only clips on the high end. I believe this is the difference you are seeing, but without a sample I can only speculate.

I don't see what point you are trying to make. I can assure you vo_gpu is not better/correct. libplacebo has it's problems, but will not find them comparing to unmaintained code of vo_gpu from years ago. Plot twist is that this tone mapping was also authored by haasn and libplacebo is evolution of it if one can describe it like so.

**Comment by pbiancardi** [comment]
> Spline contrast is exposed as `--tone-mapping-param`, contrast recovery is not exposed as an option yet; will add it soon.
> 
> As a side note, the above change and a whole pile of other related tone mapping improvements are sitting on a branch at https://code.videolan.org/videolan/libplacebo/-/merge_requests/481
> 
> I'll merge that and make the libplacebo release before updating mpv

Some of us are confused, has this made it to the MPV nightly builds yet?  Want to make sure we are looking at the right thing.

**Comment by haasn** [comment]
> > Spline contrast is exposed as `--tone-mapping-param`, contrast recovery is not exposed as an option yet; will add it soon.
> > As a side note, the above change and a whole pile of other related tone mapping improvements are sitting on a branch at https://code.videolan.org/videolan/libplacebo/-/merge_requests/481
> > I'll merge that and make the libplacebo release before updating mpv
> 
> Some of us are confused, has this made it to the MPV nightly builds yet? Want to make sure we are looking at the right thing.

By now, yes

**Comment by pbiancardi** [comment]
> > > Spline contrast is exposed as `--tone-mapping-param`, contrast recovery is not exposed as an option yet; will add it soon.
> > > As a side note, the above change and a whole pile of other related tone mapping improvements are sitting on a branch at https://code.videolan.org/videolan/libplacebo/-/merge_requests/481
> > > I'll merge that and make the libplacebo release before updating mpv
> > 
> > 
> > Some of us are confused, has this made it to the MPV nightly builds yet? Want to make sure we are looking at the right thing.
> 
> By now, yes

I compared 7.4 MPV build with pre-updates and did not notice a change?  Isn't tricubic interpolation the big one?  That is not exposed in MPV yet?

**Comment by movieperson** [comment]
First things first:  You are making great progress with these tone and gamut improvements.

However, there are still an issue that I see as we get to higher MCL's, mainly with the relationship between the 3 primary colors.  Most affected is red, then green.  At MCL 4000, red is severely clipped and green is also starting to be clipped. There also seems to be an issue with blue related to hue linearity, regardless of MCL.

I'm including some screen shots to illustrate what I see.  I hope these will help you home into additional possible improvements.

Patterns are from S&M V3 UHD disc.  Player is MPV build mpv-x86_64-20230708-git-f5eb7ea, using the following pertinent config statements:

vo=gpu-next
hwdec=auto
gpu-api=d3d11
gpu-context=d3d11
target-trc=gamma2.4
target-prim=bt.709
tone-mapping=bt.2446a # use for HDR to SDR
gamut-clipping=no
gamut-mapping-mode=perceptual
hdr-compute-peak=yes


![MCL 1000-R](https://github.com/haasn/libplacebo/assets/72896057/51e5a6c1-e4af-4e9d-9d7c-2929d63d1165)
![MCL 1000-G](https://github.com/haasn/libplacebo/assets/72896057/67e90b1e-6477-41fc-a2ee-d7b0785d6852)
![MCL 1000-B](https://github.com/haasn/libplacebo/assets/72896057/78456806-1eaa-4ff3-955a-54feb3027c4a)
![MCL 1000-patches](https://github.com/haasn/libplacebo/assets/72896057/e2991508-dd4a-4b62-9f16-53c4dae22156)
![MCL 2000 -R](https://github.com/haasn/libplacebo/assets/72896057/9252be25-f2af-49b1-b4e0-9a89265d653c)
![MCL 4000-R](https://github.com/haasn/libplacebo/assets/72896057/e522366e-01cc-456d-8b35-5ceb53475bfe)
![MCL 4000-G](https://github.com/haasn/libplacebo/assets/72896057/e6e49eb6-8429-495e-ac92-936d5e764191)
![MCL 4000-patches](https://github.com/haasn/libplacebo/assets/72896057/7f1d40da-6e2a-4d6f-8037-b7093895676e)

**Comment by danbezerra** [comment]
Haans, I am following the great improvements in the tone mapping area and the positive feedback over the JRiver forum. Do you expect to expose tricubic and tone mapping target contrast options on MPV ? If I understand correctly, most if not all of the code is part of LibPlacebo version consumed by MPV daily builds, but the users have no option yet to select them - Is this correct?

Thanks!

**Comment by haasn** [comment]
The tone mapping contrast is already exposed (it's the tone mapping parameter). I could add an option for tricubic interpolation, but it really isn't that big a deal outside of synthetic test patterns.

**Comment by danbezerra** [comment]
Thanks. I personally don't care about synthetic test patterns, but I'm sure others would like to see it so you may want to expose tricubic as well. Plus, part of your recommendation above is to also set the contrast recovery but this one is not exposed yet (AFAIK).

**Comment by haasn** [comment]
Probably will be fixed by https://code.videolan.org/videolan/libplacebo/-/merge_requests/516, which also makes the need for tricubic interpolation redundant to the point of not being worth an option (IMO).

**Comment by ThinkMcFlyThink** [comment]
> @nicko88 seems the patch I posted earlier helps a lot, I will merge it
> 
> Apart from that, I found that increasing both 'Contrast recovery' (to about 0.7) and 'Contrast' (for spline, to about 0.9) gave me a look&feel similar to madVR, just testing on the first few screenshots:
> 
> ![libplacebo-50-mod-custom](https://user-images.githubusercontent.com/1149047/249101334-2106f573-1643-4fad-b021-8ed23334b5e6.png) ![libplacebo3](https://user-images.githubusercontent.com/1149047/249102162-ca5a0a9a-2864-4567-b5d7-6595f5fa28e7.png)

Slightly off-topic, but after no luck searching I must ask here. What script is it which gives you the settings menu? Cheers.

**Comment by Hrxn** [comment]
Not a script, AFAIK, looks like the included "demo" program of libplacebo? plplay or something?
